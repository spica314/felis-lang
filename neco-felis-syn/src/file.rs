use crate::{Item, Parse, ParseError, token::Token};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct File {
    items: Vec<Item>,
}

impl File {
    pub fn items(&self) -> &[Item] {
        &self.items
    }
}

impl Parse for File {
    fn parse(tokens: &[Token], i: &mut usize) -> Result<Option<Self>, ParseError> {
        let mut k = *i;

        let mut items = vec![];
        while let Some(item) = Item::parse(tokens, &mut k)? {
            items.push(item);
        }

        let file = File { items };

        *i = k;
        Ok(Some(file))
    }
}

#[cfg(test)]
mod test {
    use crate::{FileIdGenerator, token::Token};

    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse_inductive_eq() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_eq.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_inductive_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_eq_and_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/eq_and_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }
}
