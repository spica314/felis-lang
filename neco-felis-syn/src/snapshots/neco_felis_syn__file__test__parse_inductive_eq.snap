---
source: neco-felis-syn/src/file.rs
expression: file
---
File {
    items: [
        Inductive(
            ItemInductive {
                keyword_inductive: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 1,
                    },
                    s: "inductive",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 12,
                    },
                    s: "eq",
                },
                colon: TokenColon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 15,
                    },
                },
                ty: ArrowDep(
                    TermArrowDep {
                        paren_l: TokenParenL {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 1,
                                column: 17,
                            },
                        },
                        from: TermVariable {
                            variable: TokenVariable {
                                pos: Pos {
                                    file_id: FileId(
                                        0,
                                    ),
                                    line: 1,
                                    column: 18,
                                },
                                s: "A",
                            },
                            ext: (),
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 1,
                                column: 20,
                            },
                        },
                        from_ty: Apply(
                            TermApply {
                                f: Variable(
                                    TermVariable {
                                        variable: TokenVariable {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 1,
                                                column: 22,
                                            },
                                            s: "Type",
                                        },
                                        ext: (),
                                    },
                                ),
                                args: [],
                                ext: (),
                            },
                        ),
                        paren_r: TokenParenR {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 1,
                                column: 26,
                            },
                        },
                        arrow: TokenOperator {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 1,
                                column: 28,
                            },
                            s: "->",
                        },
                        to: ArrowDep(
                            TermArrowDep {
                                paren_l: TokenParenL {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 31,
                                    },
                                },
                                from: TermVariable {
                                    variable: TokenVariable {
                                        pos: Pos {
                                            file_id: FileId(
                                                0,
                                            ),
                                            line: 1,
                                            column: 32,
                                        },
                                        s: "x",
                                    },
                                    ext: (),
                                },
                                colon: TokenColon {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 34,
                                    },
                                },
                                from_ty: Apply(
                                    TermApply {
                                        f: Variable(
                                            TermVariable {
                                                variable: TokenVariable {
                                                    pos: Pos {
                                                        file_id: FileId(
                                                            0,
                                                        ),
                                                        line: 1,
                                                        column: 36,
                                                    },
                                                    s: "A",
                                                },
                                                ext: (),
                                            },
                                        ),
                                        args: [],
                                        ext: (),
                                    },
                                ),
                                paren_r: TokenParenR {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 37,
                                    },
                                },
                                arrow: TokenOperator {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 39,
                                    },
                                    s: "->",
                                },
                                to: ArrowNodep(
                                    TermArrowNodep {
                                        from: Variable(
                                            TermVariable {
                                                variable: TokenVariable {
                                                    pos: Pos {
                                                        file_id: FileId(
                                                            0,
                                                        ),
                                                        line: 1,
                                                        column: 42,
                                                    },
                                                    s: "A",
                                                },
                                                ext: (),
                                            },
                                        ),
                                        arrow: TokenOperator {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 1,
                                                column: 44,
                                            },
                                            s: "->",
                                        },
                                        to: Apply(
                                            TermApply {
                                                f: Variable(
                                                    TermVariable {
                                                        variable: TokenVariable {
                                                            pos: Pos {
                                                                file_id: FileId(
                                                                    0,
                                                                ),
                                                                line: 1,
                                                                column: 47,
                                                            },
                                                            s: "Prop",
                                                        },
                                                        ext: (),
                                                    },
                                                ),
                                                args: [],
                                                ext: (),
                                            },
                                        ),
                                        ext: (),
                                    },
                                ),
                                ext: (),
                            },
                        ),
                        ext: (),
                    },
                ),
                brace_l: TokenBraceL {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 52,
                    },
                },
                branches: [
                    ItemInductiveBranch {
                        name: TokenVariable {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 5,
                            },
                            s: "eq_refl",
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 12,
                            },
                        },
                        ty: ArrowDep(
                            TermArrowDep {
                                paren_l: TokenParenL {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 2,
                                        column: 14,
                                    },
                                },
                                from: TermVariable {
                                    variable: TokenVariable {
                                        pos: Pos {
                                            file_id: FileId(
                                                0,
                                            ),
                                            line: 2,
                                            column: 15,
                                        },
                                        s: "A",
                                    },
                                    ext: (),
                                },
                                colon: TokenColon {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 2,
                                        column: 17,
                                    },
                                },
                                from_ty: Apply(
                                    TermApply {
                                        f: Variable(
                                            TermVariable {
                                                variable: TokenVariable {
                                                    pos: Pos {
                                                        file_id: FileId(
                                                            0,
                                                        ),
                                                        line: 2,
                                                        column: 19,
                                                    },
                                                    s: "Type",
                                                },
                                                ext: (),
                                            },
                                        ),
                                        args: [],
                                        ext: (),
                                    },
                                ),
                                paren_r: TokenParenR {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 2,
                                        column: 23,
                                    },
                                },
                                arrow: TokenOperator {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 2,
                                        column: 25,
                                    },
                                    s: "->",
                                },
                                to: ArrowDep(
                                    TermArrowDep {
                                        paren_l: TokenParenL {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 28,
                                            },
                                        },
                                        from: TermVariable {
                                            variable: TokenVariable {
                                                pos: Pos {
                                                    file_id: FileId(
                                                        0,
                                                    ),
                                                    line: 2,
                                                    column: 29,
                                                },
                                                s: "x",
                                            },
                                            ext: (),
                                        },
                                        colon: TokenColon {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 31,
                                            },
                                        },
                                        from_ty: Apply(
                                            TermApply {
                                                f: Variable(
                                                    TermVariable {
                                                        variable: TokenVariable {
                                                            pos: Pos {
                                                                file_id: FileId(
                                                                    0,
                                                                ),
                                                                line: 2,
                                                                column: 33,
                                                            },
                                                            s: "A",
                                                        },
                                                        ext: (),
                                                    },
                                                ),
                                                args: [],
                                                ext: (),
                                            },
                                        ),
                                        paren_r: TokenParenR {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 34,
                                            },
                                        },
                                        arrow: TokenOperator {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 36,
                                            },
                                            s: "->",
                                        },
                                        to: Apply(
                                            TermApply {
                                                f: Variable(
                                                    TermVariable {
                                                        variable: TokenVariable {
                                                            pos: Pos {
                                                                file_id: FileId(
                                                                    0,
                                                                ),
                                                                line: 2,
                                                                column: 39,
                                                            },
                                                            s: "eq",
                                                        },
                                                        ext: (),
                                                    },
                                                ),
                                                args: [
                                                    Variable(
                                                        TermVariable {
                                                            variable: TokenVariable {
                                                                pos: Pos {
                                                                    file_id: FileId(
                                                                        0,
                                                                    ),
                                                                    line: 2,
                                                                    column: 42,
                                                                },
                                                                s: "A",
                                                            },
                                                            ext: (),
                                                        },
                                                    ),
                                                    Variable(
                                                        TermVariable {
                                                            variable: TokenVariable {
                                                                pos: Pos {
                                                                    file_id: FileId(
                                                                        0,
                                                                    ),
                                                                    line: 2,
                                                                    column: 44,
                                                                },
                                                                s: "x",
                                                            },
                                                            ext: (),
                                                        },
                                                    ),
                                                    Variable(
                                                        TermVariable {
                                                            variable: TokenVariable {
                                                                pos: Pos {
                                                                    file_id: FileId(
                                                                        0,
                                                                    ),
                                                                    line: 2,
                                                                    column: 46,
                                                                },
                                                                s: "x",
                                                            },
                                                            ext: (),
                                                        },
                                                    ),
                                                ],
                                                ext: (),
                                            },
                                        ),
                                        ext: (),
                                    },
                                ),
                                ext: (),
                            },
                        ),
                        comma: TokenComma {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 47,
                            },
                        },
                        ext: (),
                    },
                ],
                brace_r: TokenBraceR {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 1,
                    },
                },
                ext: (),
            },
        ),
    ],
    ext: (),
}
